frame/asset-rate/src/lib.rs

```diff
@@ -0,0 +1,247 @@
+//! * **Native balance**: The balance type of the network's native currency.
+//! * **Treasury spend**: A payment from the treasury after the corresponding proposal has been
+//!   approved.
+//!
+//! ### Goals
+//!
+//! The asset-rate system in Substrate is designed to make the following possible:
+//!
+//! * Whitelisting assets other than the native currency which can be accepted for Treasury spends.
+//! * Providing a soft conversion for the balance of whitelisted assets to native.
+//! * Updating existing conversion rates.
+//!
+//! ## Interface
+//!
+//! ### Permissioned Functions
+//!
+//! * `create`: Creates a new asset conversion rate.
+//! * `remove`: Removes an existing asset conversion rate.
+//! * `update`: Overwrites an existing assert conversion rate.
+//!
+//! Please refer to the [`Call`] enum and its associated variants for documentation on each
+//! function.
+//!
+//! ### Assumptions
+//!
+//! * Conversion rates will not be used to determine the payment amount in another asset.
+//! * Conversion rates will be used to determine the tier of the spender status, e.g.
+//!   `SmallSpender`, `MediumSpender`, `BigSpender` or `Treasurer`.
+//! * All conversion rates reflect the ration of some asset to native, e.g. native = asset * rate.
+//!
+//! ## Related Modules
+//! * [`Treasury`](../treasury/index.html)
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::traits::{
+	fungible::Inspect,
+	tokens::{Balance, BalanceConversion},
+};
+use sp_runtime::{traits::Zero, FixedPointNumber, FixedPointOperand, FixedU128};
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+#[cfg(feature = "runtime-benchmarks")]
+pub mod benchmarking;
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+pub mod weights;
+
+// Type alias for `frame_system`'s account id.
+type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+// This pallet's asset id and balance type.
+type AssetIdOf<T> = <T as Config>::AssetId;
+// Generic fungible balance type.
+type BalanceOf<T> = <<T as Config>::Currency as Inspect<AccountIdOf<T>>>::Balance;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+
+		/// The runtime event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The origin permissioned to create a conversion rate for an asset.
+		type CreateOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// The origin permissioned to remove an existing conversion rate for an asset.
+		type RemoveOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// The origin permissioned to update an existiing conversion rate for an asset.
+		type UpdateOrigin: EnsureOrigin<Self::RuntimeOrigin>;
```

Comment:
I wonder though, are there strong benefits to having these as separate origins?

Reply:
IMO, at least the `UpdateOrigin` and `{Create, Remove}Origin` should be separated as updating might be handled periodically by another origin then adding/removing which most likely falls back to Governance. 

However, I believe we should provide the most flexible granularity and keep all three origins. Consider separate Gov2 track thresholds for adding and removing an asset rate as one argument. In the end, even if `CreateOrigin` and `RemoveOrigin` are exactly the same, it's one more line of configuration the particular runtime.